pkg/koordlet/statesinformer/states_nodemetric.go

```diff
@@ -485,25 +491,115 @@ func (r *nodeMetricInformer) collectNodeAggregateMetric(endTime time.Time, aggre
 	return aggregateUsages
 }
 
-func (r *nodeMetricInformer) collectPodMetric(podMeta *PodMeta, queryParam *metriccache.QueryParam) *slov1alpha1.PodMetricInfo {
+func (r *nodeMetricInformer) collectPodMetric(podMeta *PodMeta, queryParam metriccache.QueryParam) *slov1alpha1.PodMetricInfo {
 	if podMeta == nil || podMeta.Pod == nil {
 		return nil
 	}
 	podUID := string(podMeta.Pod.UID)
-	queryResult := r.metricCache.GetPodResourceMetric(&podUID, queryParam)
-	if queryResult.Error != nil {
-		klog.Warningf("get pod %v resource metric failed, error %v", podUID, queryResult.Error)
+
+	queryer, err := r.metricCache.Querier(*queryParam.Start, *queryParam.End)
+	if err != nil {
+		klog.Warningf("get query failed, error %v", err)
 		return nil
 	}
-	if queryResult.Metric == nil {
-		klog.Warningf("pod %v metric not exist", podUID)
+
+	cpuAggregateResult, err := doQuery(queryer, metriccache.PodCPUUsageMetric, metriccache.MetricPropertiesFunc.Pod(podUID))
+	if err != nil {
 		return nil
 	}
-	return &slov1alpha1.PodMetricInfo{
+	cpuUsed, err := cpuAggregateResult.Value(queryParam.Aggregate)
+	if err != nil {
+		return nil
+	}
+
+	memAggregateResult, err := doQuery(queryer, metriccache.PodMemUsageMetric, metriccache.MetricPropertiesFunc.Pod(podUID))
+	if err != nil {
+		return nil
+	}
+
+	memUsed, err := memAggregateResult.Value(queryParam.Aggregate)
+	if err != nil {
+		return nil
+	}
+	rtn := &slov1alpha1.PodMetricInfo{
 		Namespace: podMeta.Pod.Namespace,
 		Name:      podMeta.Pod.Name,
-		PodUsage:  *convertPodMetricToResourceMap(queryResult.Metric),
+		PodUsage: slov1alpha1.ResourceMap{
+			ResourceList: corev1.ResourceList{
+				corev1.ResourceCPU:    *resource.NewMilliQuantity(int64(cpuUsed*1000), resource.DecimalSI),
+				corev1.ResourceMemory: *resource.NewQuantity(int64(memUsed), resource.BinarySI),
+			},
+		},
 	}
+
+	value, exist := r.metricCache.Get(koordletutil.GPUDeviceType)
```

Comment:
This is a node-level resource and there could be many nodes without GPU. We'd better check if the GPU metric collection is necessary outside the `collectPodMetric()` and call `collectPodGPUMetric()` individually.